Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift

```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
-                buffer[0..<unusedDigits].initialize(repeating: 0) // The buffer is permitted to be partially uninitialised, but only at the end.  So we have to initialise the unused portion at the start, in principle.  Technically this probably doesn't matter given we never subsequently read this part of the buffer, but there's no way to express that such that the compiler can ensure it for us.
-            }
+        numericStringRepresentationForBinaryInteger(words: self.words, isSigned: Self.isSigned)
+    }
+}
 
-            initialisedCount = maximumDigits
-        }[(maximumDigits - actualDigits)...]
+/// Formats `words` in "Numeric string" format (https://speleotrove.com/decimal/daconvs.html)
+/// which is the required input form for certain ICU functions (e.g. `unum_formatDecimal`).
+///
+/// - Parameters:
+///   - words: The binary integer's words.
+///   - isSigned: The binary integer's signedness.
+///
+internal // In order to test big integers without big integer models.
+func numericStringRepresentationForBinaryInteger(
+words: some Collection<UInt>, isSigned: Bool) -> ArraySlice<UInt8> {
+    // Copies the words and then passes them to a non-generic, mutating, word-based algorithm.
+    withUnsafeTemporaryAllocation(of: UInt.self, capacity: words.count) { copy in
+        let count =  copy.initialize(fromContentsOf: words)
+        defer{ copy.baseAddress!.deinitialize(count: count) }
+        return numericStringRepresentationForMutableBinaryInteger(words: copy, isSigned: isSigned)
     }
-    
-    /// - Parameter bitWidth: The bit width of interest.  Must be zero or positive.
-    /// - Returns: The maximum number of decimal digits that an unsigned value of the given bit width may contain.
-    internal static func maximumDecimalDigitsForUnsigned(bitWidth: Int) -> Int {
-        guard 0 < bitWidth else { return 0 }
-        guard 1 != bitWidth else { return 1 } // Algorithm below only works for bit widths of _two_ and above.
+}
 
-        return Int((Double(bitWidth) * log10(2)).rounded(.up)) + 1 // https://www.exploringbinary.com/number-of-decimal-digits-in-a-binary-integer
+/// Formats `words` in "Numeric string" format (https://speleotrove.com/decimal/daconvs.html)
+/// which is the required input form for certain ICU functions (e.g. `unum_formatDecimal`).
+///
+/// - Parameters:
+///   - words: The binary integer's mutable words.
+///   - isSigned: The binary integer's signedness.
+///
+/// This method consumes the `words` such that the buffer is filled with zeros when it returns.
+///
+private func numericStringRepresentationForMutableBinaryInteger(
+words: UnsafeMutableBufferPointer<UInt>, isSigned: Bool) -> ArraySlice<UInt8> {
+    //  Note that negative values are in two's complement form.
+    let isLessThanZero = isSigned && Int(bitPattern: words.last ?? 0) < 0
+    //  The magnitude is formed when the words represent a negative value.
+    if  isLessThanZero {
+        formTwosComplementForBinaryInteger(words: words)
```

Comment:
While it's possible to move this transformation to the `BinaryInteger` extension, it's only viable if `@inlinable` because it won't be specialized otherwise. The unspecialized performance penalty of evaluating something like `Self.isSigned && self < Self.zero` is quite apparent for `Int256`, but it can also seen in `Int`. If you are all OK with `@inlinable`, removing this method and calling `magnitude.words` is the way to go about it.